var/spack/repos/builtin/packages/singularity/package.py

```diff
@@ -5,29 +5,124 @@
+        with working_dir(self.build_directory):
+            configure = Executable('./mconfig --prefix=%s' % prefix)
+            configure()
+
+    # Set these for use by MakefilePackage's default build/install methods.
+    build_targets = ['-C', 'builddir', 'parallel=False']
+    install_targets = ['install', '-C', 'builddir', 'parallel=False']
+
+    def setup_environment(self, spack_env, run_env):
+        # Point GOPATH at the top of the staging dir for the build
+        # step.
+        spack_env.prepend_path('GOPATH', self.stage.path)
+
+    # `singularity` has a fixed path where it will look for
+    # mksquashfs.  If it lives somewhere else you need to specify the
+    # full path in the config file.  Fix the config file after it's
+    # installed.
+    @run_after('install')
+    def fix_mksquashfs_path(self):
+        prefix = self.spec.prefix
+        squash_path = join_path(self.spec['squashfs'].prefix.bin, 'mksquashfs')
+        filter_file(r'^# mksquashfs path =',
+                    'mksquashfs path = {0}'.format(squash_path),
+                    join_path(prefix.etc, 'singularity', 'singularity.conf'))
+
+    #
+    # Assemble a script that fixes the ownership and permissions of several
+    # key files, install it, and tty.warn() the user.
+    # HEADSUP: https://github.com/spack/spack/pull/10412.
+    #
+    def perm_script(self):
+        return 'spack_perms_fix.sh'
+
+    def perm_script_tmpl(self):
+        return "{0}.j2".format(self.perm_script())
+
+    def perm_script_path(self):
+        return join_path(self.spec.prefix.bin, self.perm_script())
+
+    def _build_script(self, filename, variable_data):
+        with open(filename, 'w') as f:
+            env = spack.tengine.make_environment(dirs=self.package_dir)
+            t = env.get_template(self.perm_script_tmpl())
+            f.write(t.render(variable_data))
+
+    @run_after('install')
+    def build_perms_script(self):
+        script = self.perm_script_path()
+        chown_files = ['libexec/singularity/bin/starter-suid',
+                       'etc/singularity/singularity.conf',
+                       'etc/singularity/capability.json',
+                       'etc/singularity/ecl.toml']
+        setuid_files = ['libexec/singularity/bin/starter-suid']
+        self._build_script(script, {'prefix': self.spec.prefix,
+                                    'chown_files': chown_files,
+                                    'setuid_files': setuid_files})
+        chmod = which('chmod')
+        chmod('555', script)
+
+    # Until tty output works better from build steps, this ends up in
+    # the build log.  See https://github.com/spack/spack/pull/10412.
+    @run_after('install')
+    def caveats(self):
+        tty.warn("""
+        For full functionality, you'll need to chown and chmod some files
+        after installing the package.  This has security implications.
+        See: https://singularity.lbl.gov/docs-security for details.
```

Comment:
@vsoch -- Do you know of a version of this doc for the sylabs/3 releases?  I can't find any mention of setuid bits in the new docs.

Reply:
To allow setuid? -> https://www.sylabs.io/guides/3.1/user-guide/security_options.html?highlight=setuid#security-related-action-options

Reply:
I don't think there is as substantial a security page as there used to be - if you took a look at that link and it's not sufficient, we can open an issue and ask for -vvvvv

Reply:
I believe that the link you suggested is a discussion of whether/how to configure singularity to allow it to run things that are `setuid` in containers.

I think that the older page discussed the issues around ownership and mode of files that get installed, which is a separate issue.

It's out of date, but this bit (https://singularity.lbl.gov/docs-security#where-are-the-singularity-privileged-components) is germane to us, plus a discussion of the various files that need to be root owned.

Reply:
Maybe in the admin docs? https://www.sylabs.io/guides/3.1/admin-guide/admin_quickstart.html#configuration

I agree it's not as substantial as before.

Reply:
https://github.com/sylabs/singularity-admindocs/issues/39